### PR TITLE
Implement lazy module initialization/compilation

### PR DESCRIPTION
**Pull request by linehill**
Compile SPIR-V modules and initialize them just before they are needed instead of doing so for all of the modules at once at the start of the program to reduce start-up time of the HIP applications. For a case in #283 the start-up and compilation time reduces to ~45 seconds from the previous ~40 minutes (as the case only launches one kernel).

Issues with the CHIP runtime uninitialization were discovered while making this patch for the following cases:

```
Unit_hipMemcpyWithStream_MultiThread
Unit_hipStreamPerThread_MultiThread
Unit_hipStreamPerThread_DeviceReset_1
```

All of them features a main thread, which spawns worker threads, and then exits the `main()` and uninitializes the CHIP runtime while the work threads are still running. The failures could have popped up earlier. This patch just makes them appear more often.

### CODE REVIEW

**Review by pjaaskel — CHANGES_REQUESTED** [review]
Great stuff, excellent speedups. Only nitpitcks here.

File: src/CHIPSourceRegister.cc [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+  if (SrcMod->FinalizedSource_.size()) // Already processed source. Return it.
+    return SrcMod;
+
+  // TODO: Optimization: Try to split the original large source module
+  //       into smaller independent ones (is possible) for reducing
+  //       compilation time in the backend.
+
+  bool Success =
+      filterSPIRV(SrcMod->OriginalSource_.data(),
```
[the remark this answers is not in the document]
Comment (linehill) [reply src/CHIPSourceRegister.cc]: FilterSPIRV is a step in the "finalization". I visioned bringing `CHIPModule::consumeSPIRV()` here but not in this patch for limiting the scope of the review.

File: src/CHIPSourceRegister.hh [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  std::list<CHIPSourceFunction> Kernels;
+  std::list<CHIPSourceVariable> Variables;
+
+  std::string_view getSourceCode() const {
+    assert(FinalizedSource_.size() && "Has not finalized yet!");
+    return FinalizedSource_;
+  }
+};
+
+class CHIPSourceRegister {
```
[the remark this answers is not in the document]
Comment (linehill) [reply src/CHIPSourceRegister.hh]: Renamed to SPVRegister. Did the same for the other CHIP* classes.

File: src/CHIPBindings.cc [hunk 3]
```diff
@@ -3992,13 +3976,14 @@ __hipRegisterVar(void **Data,
   assert(std::string(HostName) == std::string(DeviceName));
 
   CHIP_TRY
-  CHIPInitialize();
+  // NOTE: CHIP backend initialization is undesired here.
```
[the remark this answers is not in the document]
Comment (linehill) [reply src/CHIPBindings.cc]: Added rationale.

File: src/CHIPSourceRegister.hh [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+struct HostPtr {
+  const void *Value;
+
+  template <typename T> explicit HostPtr(T *TheValue) : Value(TheValue) {}
+
+  operator const void *() const { return Value; }
+};
+
+/// Represents a function or global variable in a source module.
+struct CHIPSourceEntity {
```
[the remark this answers is not in the document]
Comment (linehill) [reply src/CHIPSourceRegister.hh]: Renamed to SPVGlobalObject.

File: CMakeLists.txt [hunk 5]
```diff
@@ -315,6 +317,14 @@ list(APPEND FAILING_TESTS "Unit_funnelshift")
...
 list(APPEND FAILING_TESTS "Unit_syncthreads_count")
 list(APPEND FAILING_TESTS "Unit_syncthreads_or")
+# The following three cases have the same scenario: Main thread spawn a set of
+# worker threads and then the main thread return from main without waiting
+# the worker threads to finish. The worker threads segfaults while accessing
+# destructed CHIP-SPV. Tests may sometimes pass but the chances are low after
+# lazy module initialization patch.
+list(APPEND FAILING_TESTS "Unit_hipMemcpyWithStream_MultiThread")
+list(APPEND FAILING_TESTS "Unit_hipStreamPerThread_MultiThread")
+list(APPEND FAILING_TESTS "Unit_hipStreamPerThread_DeviceReset_1")
```
Comment (pvelesko) [inline CMakeLists.txt:327]: I just ran `Unit_hipMemcpyWithStream_MultiThread` 500 times on main and it doesn't crash. It crashed on this branch quite quickly so the failure seems to be introduced in this patch. We should resolve before merging.

Comment (pjaaskel) [reply CMakeLists.txt:327]: When talking about race conditions, major optimizations in execution speed can cause races to clash faster. Here the speedup is quite drastic so it's likely the case here.

Comment (linehill) [reply CMakeLists.txt]: Reverted this change.

**Comment by linehill** [comment]
* Fixed a timing sensitive multi-threading issue (unrelated to the new feature in this patch). See the “Fix call configurations got mixed up” commit for the issue and solution.
* Restored initialization of CHIP-SPV runtime at __hipRegisterFatBinary() call to avert test failures in some test cases (e.g. Unit_hipStreamPerThread_DeviceReset_1).

**Comment by pvelesko** [comment]
> Fixed a timing sensitive multi-threading issue (unrelated to the new feature in this patch). See the “Fix call configurations got mixed up” commit for the issue and solution. 

That's great! Thanks for resolving it. 

I'll review and merge today

**Review by pvelesko — APPROVED** [review]

**Comment by linehill** [comment]
Thanks, @pjaaskel and @pjaaskel.

**Pull request merged**